Calling std::random or rand_s causes crash on Windows 2000, because they never added a fallback handler to use
the (less secure, but Win2K compatible) CryptGenRandom when RtlGenRandom fails. Fix and add fallback.

diff --git a/mingw-w64-crt/secapi/rand_s.c b/mingw-w64-crt/secapi/rand_s.c
index 52700ba..c28f56a 100644
--- a/mingw-w64-crt/secapi/rand_s.c
+++ b/mingw-w64-crt/secapi/rand_s.c
@@ -6,10 +6,36 @@
 #include <msvcrt.h>
 
 static BOOLEAN (WINAPI *pRtlGenRandom)(void*,ULONG);
+static BOOLEAN (WINAPI *pCryptAcquireContext)(HCRYPTPROV*,LPCWSTR,LPCWSTR,DWORD,DWORD);
+static BOOLEAN (WINAPI *pCryptGenRandom)(HCRYPTPROV,DWORD,BYTE*);
+static BOOLEAN (WINAPI *pCryptReleaseContext)(HCRYPTPROV,DWORD);
 
 static errno_t mingw_rand_s(unsigned int *pval)
 {
-    return !pval || !pRtlGenRandom || !pRtlGenRandom(pval, sizeof(*pval)) ? EINVAL : 0;
+    if (!pval) {
+        return EINVAL;
+    }
+    if (pRtlGenRandom) {
+        return !pRtlGenRandom(pval, sizeof(*pval)) ? EINVAL : 0;
+    } else {
+        if (!pCryptGenRandom || !pCryptAcquireContext || !pCryptReleaseContext) {
+            return EINVAL;
+        }
+        unsigned int kSeed = 0;
+        BOOL bCryptGen = FALSE;
+        HCRYPTPROV hProvider = 0;
+        // CryptGenRandom is available on Windows 2000+ (unlike RtlGenRandom/rand_s which requires XP+)
+        if (pCryptAcquireContext(&hProvider, NULL, NULL, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT)) {
+            bCryptGen = pCryptGenRandom(hProvider, sizeof(*pval), (BYTE*)(&kSeed));
+            pCryptReleaseContext(hProvider, 0);
+        }
+        if (!bCryptGen || kSeed == 0) {
+            return EINVAL;
+        } else {
+            *pval = kSeed;
+            return 0;
+        }
+    }
 }
 
 static errno_t __cdecl init_rand_s(unsigned int*);
@@ -30,6 +56,12 @@ static errno_t __cdecl init_rand_s(unsigned int *val)
     if(!func) {
         func = mingw_rand_s;
         pRtlGenRandom = (void*)GetProcAddress(LoadLibraryW(L"advapi32.dll"), "SystemFunction036");
+        if (!pRtlGenRandom) {
+            // CryptGenRandom is available on Windows 2000+
+            pCryptGenRandom = (void*)GetProcAddress(LoadLibraryW(L"advapi32.dll"), "CryptGenRandom");
+            pCryptAcquireContext = (void*)GetProcAddress(LoadLibraryW(L"advapi32.dll"), "CryptAcquireContextW");
+            pCryptReleaseContext = (void*)GetProcAddress(LoadLibraryW(L"advapi32.dll"), "CryptReleaseContext");
+        }
     }
 
     return (__MINGW_IMP_SYMBOL(rand_s) = func)(val);
